### PR TITLE
Add support of conf file to pass arguments and options

### PR DESCRIPTION
**Pull request by vadmeste**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
Allow minio server to take inputs via config.yaml

## Motivation and Context
provides more flexibility and centralizes the
minio configuration away from /etc/default/minio
on bare-metal to environment values in k8s.

Both of them can be converted to a single file,
allowing k8s to save the entire MinIO config in
a configMap/k8s secret. 

## How to test this PR?
```
minio server --config config.yaml
```

```yaml
version: v1
address: ':9000'
console-address: ':9001'
certs-dir: '/home/user/.minio/certs/'
pools: # Specify the nodes and drives with pools
  -
        - 'https://server-example-pool1:9000/mnt/disk{1...4}/'
        - 'https://server{1...2}-pool1:9000/mnt/disk{1...4}/'
        - 'https://server3-pool1:9000/mnt/disk{1...4}/'
        - 'https://server4-pool1:9000/mnt/disk{1...4}/'
  -
        - 'https://server-example-pool2:9000/mnt/disk{1...4}/'
        - 'https://server{1...2}-pool2:9000/mnt/disk{1...4}/'
        - 'https://server3-pool2:9000/mnt/disk{1...4}/'
        - 'https://server4-pool2:9000/mnt/disk{1...4}/'

options:
  ftp: # settings for MinIO to act as an FTP server
        address: ':8021'
        passive-port-range: '30000-40000'
  sftp: # settings for MinIO to act as an FTP server
        address: ':8022'
        ssh-private-key: '/home/user/.ssh/id_rsa'
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [x] Unit tests added/updated
- [x] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/endpoint.go [hunk 1]
```diff
@@ -805,21 +805,25 @@ func (p PoolEndpointList) UpdateIsLocal() error {
 
 // CreatePoolEndpoints creates a list of endpoints per pool, resolves their relevant hostnames and
 // discovers those are local or remote.
-func CreatePoolEndpoints(serverAddr string, poolArgs ...[][]string) ([]Endpoints, SetupType, error) {
+func CreatePoolEndpoints(serverAddr string, poolsLayout ...poolDisksLayout) ([]Endpoints, SetupType, error) {
 	var setupType SetupType
 
+	if len(poolsLayout) == 0 || len(poolsLayout[0].layout) == 0 || len(poolsLayout[0].layout[0]) == 0 || len(poolsLayout[0].layout[0][0]) == 0 {
```
Comment (harshavardhana) [inline cmd/endpoint.go]: You can convert this into a function

File: cmd/endpoint.go [hunk 2]
```diff
@@ -805,21 +805,25 @@ func (p PoolEndpointList) UpdateIsLocal() error {
...
 
 	_, serverAddrPort := mustSplitHostPort(serverAddr)
 
-	poolEndpoints := make(PoolEndpointList, len(poolArgs))
+	poolEndpoints := make(PoolEndpointList, len(poolsLayout))
 
 	// For single arg, return single drive EC setup.
-	if len(poolArgs) == 1 && len(poolArgs[0]) == 1 && len(poolArgs[0][0]) == 1 && len(poolArgs[0][0][0]) == 1 {
-		endpoint, err := NewEndpoint(poolArgs[0][0][0])
+	if len(poolsLayout) == 1 && len(poolsLayout[0].layout) == 1 && len(poolsLayout[0].layout[0]) == 1 {
```
Comment (harshavardhana) [inline cmd/endpoint.go]: Again a function perhaps? Since now there is a proper data structure

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/server-main.go [hunk 3]
```diff
@@ -51,10 +51,17 @@ import (
...
 	"golang.org/x/exp/slices"
+	"gopkg.in/yaml.v2"
 )
 
 // ServerFlags - server command specific flags
 var ServerFlags = []cli.Flag{
+	cli.StringFlag{
+		Name:   "config-file",
+		Value:  "",
+		Usage:  "use an alternate configuration file",
```
Comment (harshavardhana) [inline cmd/server-main.go]:
```suggestion
		Usage:  "specify server configuration via YAML configuration",
```

File: cmd/server-main.go [hunk 4]
```diff
@@ -51,10 +51,17 @@ import (
 	"github.com/minio/pkg/v2/certs"
 	"github.com/minio/pkg/v2/env"
 	"golang.org/x/exp/slices"
+	"gopkg.in/yaml.v2"
 )
 
 // ServerFlags - server command specific flags
 var ServerFlags = []cli.Flag{
+	cli.StringFlag{
+		Name:   "config-file",
```
Comment (harshavardhana) [inline cmd/server-main.go]:
```suggestion
		Name:   "config",
```

File: cmd/server-main.go [hunk 5]
```diff
@@ -51,10 +51,17 @@ import (
...
+		Value:  "",
+		Usage:  "use an alternate configuration file",
+		EnvVar: "MINIO_CONFIG_FILE",
```
Comment (harshavardhana) [inline cmd/server-main.go]:
```suggestion
		EnvVar: "MINIO_CONFIG",
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/server-main.go [hunk 6]
```diff
@@ -226,9 +233,59 @@ func serverCmdArgs(ctx *cli.Context) []string {
...
+}
+
+type configFileMain struct {
+	Version     string         `yaml:"version"`
+	Addr        *string        `yaml:"address"`
+	ConsoleAddr *string        `yaml:"console-address"`
+	CertsDir    *string        `yaml:"certs-dir"`
+	Pools       [][]string     `yaml:"pools"`
+	Options     configFileOpts `yaml:"options"`
+}
```
Comment (harshavardhana) [inline cmd/server-main.go]: Also move this struct to `internal/config/server.go`

```go

// ServerConfig represents server command line options, server arguments
// and server config all in a single data structure.
type ServerConfig struct {
	Version        string     `yaml:"version"`
	Address        string     `yaml:"address"`
	ConsoleAddress string     `yaml:"console-address"`
	CertsDir       string     `yaml:"certs-dir"`
	Pools          [][]string `yaml:"pools"`
	Options        struct {
		Site struct {
			Name    string `yaml:"name"`
			Region  string `yaml:"region"`
			Comment string `yaml:"comment"`
		} `yaml:"site"`
		Ftp struct {
			Address          string `yaml:"address"`
			PassivePortRange string `yaml:"passive-port-range"`
		} `yaml:"ftp"`
		Sftp struct {
			Address       string `yaml:"address"`
			SSHPrivateKey string `yaml:"ssh-private-key"`
		} `yaml:"sftp"`
		Subnet struct {
			License string `yaml:"license"`
			APIKey  string `yaml:"api_key"`
			Proxy   string `yaml:"proxy"`
		} `yaml:"subnet"`
		Callhome struct {
			Frequency string `yaml:"frequency"`
		} `yaml:"callhome"`
		API struct {
			RequestsMax                 int    `yaml:"requests_max"`
			RequestsDeadline            string `yaml:"requests_deadline"`
			ClusterDeadline             string `yaml:"cluster_deadline"`
			CorsAllowOrigin             string `yaml:"cors_allow_origin"`
			RemoteTransportDeadline     string `yaml:"remote_transport_deadline"`
			ListQuorum                  string `yaml:"list_quorum"`
			ReplicationPriority         string `yaml:"replication_priority"`
			TransitionWorkers           string `yaml:"transition_workers"`
			StaleUploadsCleanupInterval string `yaml:"stale_uploads_cleanup_interval"`
			StaleUploadsExpiry          string `yaml:"stale_uploads_expiry"`
			DeleteCleanupInterval       string `yaml:"delete_cleanup_interval"`
			Odirect                     string `yaml:"odirect"`
			GzipObjects                 string `yaml:"gzip_objects"`
			RootAccess                  string `yaml:"root_access"`
			SyncEvents                  string `yaml:"sync_events"`
		} `yaml:"api"`
		Scanner struct {
			Speed string `yaml:"speed"`
		} `yaml:"scanner"`
		Compression struct {
			AllowEncryption string `yaml:"allow_encryption"`
			Extensions      string `yaml:"extensions"`
			MimeTypes       string `yaml:"mime_types"`
		} `yaml:"compression"`
		StorageClass struct {
			Standard string `yaml:"standard"`
			Rrs      string `yaml:"rrs"`
		} `yaml:"storage_class"`
		Heal struct {
			Bitrotscan   string `yaml:"bitrotscan"`
			MaxSleep     string `yaml:"max_sleep"`
			MaxIo        string `yaml:"max_io"`
			DriveWorkers string `yaml:"drive_workers"`
		} `yaml:"heal"`
		Etcd struct {
			Endpoints     string `yaml:"endpoints"`
			PathPrefix    string `yaml:"path_prefix"`
			CorednsPath   string `yaml:"coredns_path"`
			ClientCert    string `yaml:"client_cert"`
			ClientCertKey string `yaml:"client_cert_key"`
		} `yaml:"etcd"`
		IdentityTLS struct {
			SkipVerify string `yaml:"skip_verify"`
		} `yaml:"identity_tls"`
		IdentityPlugin struct {
			URL        string `yaml:"url"`
			AuthToken  string `yaml:"auth_token"`
			RolePolicy string `yaml:"role_policy"`
			RoleID     string `yaml:"role_id"`
		} `yaml:"identity_plugin"`
		PolicyPlugin struct {
			URL         string `yaml:"url"`
			AuthToken   string `yaml:"auth_token"`
			EnableHTTP2 string `yaml:"enable_http2"`
		} `yaml:"policy_plugin"`
		IdentityOpenid []struct {
			Name               string `yaml:"name"`
			Enable             string `yaml:"enable"`
			DisplayName        string `yaml:"display_name"`
			ConfigURL          string `yaml:"config_url"`
			ClientID           string `yaml:"client_id"`
			ClientSecret       string `yaml:"client_secret"`
			ClaimName          string `yaml:"claim_name"`
			ClaimUserinfo      string `yaml:"claim_userinfo"`
			RolePolicy         string `yaml:"role_policy"`
			ClaimPrefix        string `yaml:"claim_prefix"`
			RedirectURI        string `yaml:"redirect_uri"`
			RedirectURIDynamic string `yaml:"redirect_uri_dynamic"`
			Scopes             string `yaml:"scopes"`
			Vendor             string `yaml:"vendor"`
			KeycloakRealm      string `yaml:"keycloak_realm"`
			KeycloakAdminURL   string `yaml:"keycloak_admin_url"`
		} `yaml:"identity_openid"`
		IdentityLdap []struct {
			Name               string `yaml:"name"`
			Enable             string `yaml:"enable"`
			ServerAddr         string `yaml:"server_addr"`
			SrvRecordName      string `yaml:"srv_record_name"`
			UserDnSearchBaseDn string `yaml:"user_dn_search_base_dn"`
			UserDnSearchFilter string `yaml:"user_dn_search_filter"`
			GroupSearchFilter  string `yaml:"group_search_filter"`
			GroupSearchBaseDn  string `yaml:"group_search_base_dn"`
			TLSSkipVerify      string `yaml:"tls_skip_verify"`
			ServerInsecure     string `yaml:"server_insecure"`
			ServerStarttls     string `yaml:"server_starttls"`
			LookupBindDn       string `yaml:"lookup_bind_dn"`
			LookupBindPassword string `yaml:"lookup_bind_password"`
		} `yaml:"identity_ldap"`
		LoggerTerminal struct {
			Quiet     bool `yaml:"quiet"`
			Anonymous bool `yaml:"anonymous"`
			JSON      bool `yaml:"json"`
		} `yaml:"logger_terminal"`
		LoggerWebhook []struct {
			Name       string `yaml:"name"`
			Enable     string `yaml:"enable"`
			Endpoint   string `yaml:"endpoint"`
			AuthToken  string `yaml:"auth_token"`
			ClientCert string `yaml:"client_cert"`
			ClientKey  string `yaml:"client_key"`
			Proxy      string `yaml:"proxy"`
			QueueSize  string `yaml:"queue_size"`
			QueueDir   string `yaml:"queue_dir"`
		} `yaml:"logger_webhook"`
		AuditKafka []struct {
			Name          string `yaml:"name"`
			Enable        string `yaml:"enable"`
			Topic         string `yaml:"topic"`
			Brokers       string `yaml:"brokers"`
			SaslUsername  string `yaml:"sasl_username"`
			SaslPassword  string `yaml:"sasl_password"`
			SaslMechanism string `yaml:"sasl_mechanism"`
			ClientTLSCert string `yaml:"client_tls_cert"`
			ClientTLSKey  string `yaml:"client_tls_key"`
			TLSClientAuth int    `yaml:"tls_client_auth"`
			Sasl          string `yaml:"sasl"`
			TLS           string `yaml:"tls"`
			TLSSkipVerify string `yaml:"tls_skip_verify"`
			Version       string `yaml:"version"`
			QueueSize     int    `yaml:"queue_size"`
			QueueDir      string `yaml:"queue_dir"`
		} `yaml:"audit_kafka"`
		NotifyWebhook []struct {
			Name       string `yaml:"name"`
			Enable     string `yaml:"enable"`
			Endpoint   string `yaml:"endpoint"`
			AuthToken  string `yaml:"auth_token"`
			QueueLimit int    `yaml:"queue_limit"`
			QueueDir   string `yaml:"queue_dir"`
			ClientCert string `yaml:"client_cert"`
			ClientKey  string `yaml:"client_key"`
		} `yaml:"notify_webhook"`
		NotifyAmqp []struct {
			Name              string `yaml:"name"`
			Enable            string `yaml:"enable"`
			URL               string `yaml:"url"`
			Exchange          string `yaml:"exchange"`
			ExchangeType      string `yaml:"exchange_type"`
			RoutingKey        string `yaml:"routing_key"`
			Mandatory         string `yaml:"mandatory"`
			Durable           string `yaml:"durable"`
			NoWait            string `yaml:"no_wait"`
			Internal          string `yaml:"internal"`
			AutoDeleted       string `yaml:"auto_deleted"`
			DeliveryMode      string `yaml:"delivery_mode"`
			PublisherConfirms string `yaml:"publisher_confirms"`
			QueueLimit        int    `yaml:"queue_limit"`
			QueueDir          string `yaml:"queue_dir"`
		} `yaml:"notify_amqp"`
		NotifyKafka []struct {
			Name          string `yaml:"name"`
			Enable        string `yaml:"enable"`
			Topic         string `yaml:"topic"`
			Brokers       string `yaml:"brokers"`
			SaslUsername  string `yaml:"sasl_username"`
			SaslPassword  string `yaml:"sasl_password"`
			SaslMechanism string `yaml:"sasl_mechanism"`
			ClientTLSCert string `yaml:"client_tls_cert"`
			ClientTLSKey  string `yaml:"client_tls_key"`
			TLSClientAuth int    `yaml:"tls_client_auth"`
			Sasl          string `yaml:"sasl"`
			TLS           string `yaml:"tls"`
			TLSSkipVerify string `yaml:"tls_skip_verify"`
			QueueLimit    int    `yaml:"queue_limit"`
			QueueDir      string `yaml:"queue_dir"`
			Version       string `yaml:"version"`
			BatchSize     int    `yaml:"batch_size"`
		} `yaml:"notify_kafka"`
		NotifyMqtt []struct {
			Name              string `yaml:"name"`
			Enable            string `yaml:"enable"`
			Broker            string `yaml:"broker"`
			Topic             string `yaml:"topic"`
			Password          string `yaml:"password"`
			Username          string `yaml:"username"`
			Qos               int    `yaml:"qos"`
			KeepAliveInterval string `yaml:"keep_alive_interval"`
			ReconnectInterval string `yaml:"reconnect_interval"`
			QueueDir          string `yaml:"queue_dir"`
			QueueLimit        int    `yaml:"queue_limit"`
		} `yaml:"notify_mqtt"`
		NotifyNats []struct {
			Name                        string `yaml:"name"`
			Enable                      string `yaml:"enable"`
			Address                     string `yaml:"address"`
			Subject                     string `yaml:"subject"`
			Username                    string `yaml:"username"`
			Password                    string `yaml:"password"`
			Token                       string `yaml:"token"`
			TLS                         string `yaml:"tls"`
			TLSSkipVerify               string `yaml:"tls_skip_verify"`
			CertAuthority               string `yaml:"cert_authority"`
			ClientCert                  string `yaml:"client_cert"`
			ClientKey                   string `yaml:"client_key"`
			PingInterval                int    `yaml:"ping_interval"`
			Jetstream                   string `yaml:"jetstream"`
			Streaming                   string `yaml:"streaming"`
			StreamingAsync              string `yaml:"streaming_async"`
			StreamingMaxPubAcksInFlight int    `yaml:"streaming_max_pub_acks_in_flight"`
			StreamingClusterID          string `yaml:"streaming_cluster_id"`
			QueueDir                    string `yaml:"queue_dir"`
			QueueLimit                  int    `yaml:"queue_limit"`
		} `yaml:"notify_nats"`
		NotifyNsq []struct {
			Name          string `yaml:"name"`
			Enable        string `yaml:"enable"`
			NsqdAddress   string `yaml:"nsqd_address"`
			Topic         string `yaml:"topic"`
			TLS           string `yaml:"tls"`
			TLSSkipVerify string `yaml:"tls_skip_verify"`
			QueueDir      string `yaml:"queue_dir"`
			QueueLimit    int    `yaml:"queue_limit"`
		} `yaml:"notify_nsq"`
		NotifyMysql []struct {
			Name               string `yaml:"name"`
			Enable             string `yaml:"enable"`
			Format             string `yaml:"format"`
			DsnString          string `yaml:"dsn_string"`
			Table              string `yaml:"table"`
			QueueDir           string `yaml:"queue_dir"`
			QueueLimit         int    `yaml:"queue_limit"`
			MaxOpenConnections int    `yaml:"max_open_connections"`
		} `yaml:"notify_mysql"`
		NotifyPostgres []struct {
			Name               string `yaml:"name"`
			Enable             string `yaml:"enable"`
			Format             string `yaml:"format"`
			ConnectionString   string `yaml:"connection_string"`
			Table              string `yaml:"table"`
			QueueDir           string `yaml:"queue_dir"`
			QueueLimit         int    `yaml:"queue_limit"`
			MaxOpenConnections int    `yaml:"max_open_connections"`
		} `yaml:"notify_postgres"`
		NotifyElasticsearch []struct {
			Name       string `yaml:"name"`
			Enable     string `yaml:"enable"`
			URL        string `yaml:"url"`
			Format     string `yaml:"format"`
			Index      string `yaml:"index"`
			QueueDir   string `yaml:"queue_dir"`
			QueueLimit int    `yaml:"queue_limit"`
			Username   string `yaml:"username"`
			Password   string `yaml:"password"`
		} `yaml:"notify_elasticsearch"`
		NotifyRedis []struct {
			Name       string `yaml:"name"`
			Enable     string `yaml:"enable"`
			Format     string `yaml:"format"`
			Address    string `yaml:"address"`
			Key        string `yaml:"key"`
			Password   string `yaml:"password"`
			QueueDir   string `yaml:"queue_dir"`
			QueueLimit int    `yaml:"queue_limit"`
		} `yaml:"notify_redis"`
		LambdaWebhook []struct {
			Name       string `yaml:"name"`
			Enable     string `yaml:"enable"`
			Endpoint   string `yaml:"endpoint"`
			AuthToken  string `yaml:"auth_token"`
			ClientCert string `yaml:"client_cert"`
			ClientKey  string `yaml:"client_key"`
		} `yaml:"lambda_webhook"`
	} `yaml:"options"`
}
```

This is the final data structure that is expected in future.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Please add documentation regarding this new approach in `docs/distributed/CONFIG.md` 

We need to also provide an explanation of how this can be used, and what is the expected behavior basically what is mandatory etc.

**Review by harshavardhana — APPROVED** [review]

File: docs/distributed/CONFIG.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
+## MinIO configuration YAML
+
+MinIO now supports starting the server arguments and configuration via a YAML configuration file. This YAML configuration describes everything that can be configured in a MinIO setup, such as '--address', '--console-address' and command line arguments for the MinIO server.
```
Comment (shtripat) [inline docs/distributed/CONFIG.md:3]: I think better sentence could be `MinIO now supports starting the server using both command line arguments and configuration via a YAML configuration file` , just to be clear.

**Review by shtripat — COMMENTED** [review]
LGTM. Verified and love this feature <3.
A minor comment.

**Review by shtripat — APPROVED** [review]

File: docs/distributed/CONFIG.md (hunk 7 above)
Comment (harshavardhana) [reply docs/distributed/CONFIG.md:3]: Lets get @ravindk89's help in cleaning this up.

Comment (harshavardhana) [reply docs/distributed/CONFIG.md:3]: Will merge the PR.

**Review by Praveenrajmani — COMMENTED** [review]
tested and LGTM

**Pull request merged**